libraries/service/src/main/java/org/holochain/androidserviceruntime/service/HolochainService.kt

```diff
@@ -249,48 +350,176 @@ class HolochainService : Service() {
+    ): String {
+        Log.d(
+            logTag,
+            "showAppAuthorizationNotification clientPackageName=$clientPackageName installedAppId=$installedAppId",
+        )
+
+        // Store app authorization request
+        val putResponse = appAuthorizationRequests.put(AppAuthorizationRequest(clientPackageName, installedAppId))
+
+        // Create action intents, including id of app authorization request
+        // so we can retrieve it later
+        val approveIntent =
+            Intent(this, HolochainService::class.java).apply {
+                action = ACTION_APPROVE_APP_AUTHORIZATION
+                putExtra("requestId", putResponse.uuid)
+            }
+        val approvePendingIntent =
+            PendingIntent.getService(
+                this,
+                2,
+                approveIntent,
+                PendingIntent.FLAG_UPDATE_CURRENT or PendingIntent.FLAG_IMMUTABLE,
+            )
+
+        val denyIntent =
+            Intent(this, HolochainService::class.java).apply {
+                action = ACTION_DENY_APP_AUTHORIZATION
+                putExtra("requestId", putResponse.uuid)
+            }
+        val denyPendingIntent =
+            PendingIntent.getService(
+                this,
+                3,
+                denyIntent,
+                PendingIntent.FLAG_UPDATE_CURRENT or PendingIntent.FLAG_IMMUTABLE,
+            )
+
+        // Build the notification
+        val notification =
+            NotificationCompat
+                .Builder(this, "HolochainServiceChannel")
+                .setContentTitle("Authorize App Connection")
+                .setStyle(
+                    NotificationCompat
+                        .BigTextStyle()
+                        .bigText(
+                            "The app \"$clientPackageName\" wants to connect to the Holochain app: \"$installedAppId\".\n\nDo you want to allow this connection?",
+                        ),
+                ).setSmallIcon(R.drawable.notification_icon_sm)
+                .setPriority(NotificationCompat.PRIORITY_HIGH)
+                .setAutoCancel(true)
+                .addAction(R.drawable.ic_baseline_check, "Approve", approvePendingIntent)
+                .addAction(R.drawable.ic_outline_cancel, "Deny", denyPendingIntent)
+                .build()
+
+        // Show the notification
+        if (checkNotificationPermission()) {
+            NotificationManagerCompat.from(this).notify(NOTIFICATION_CHANNEL_ID_APP_AUTHORIZATION, putResponse.notificationId, notification)
+        } else {
+            Log.w(logTag, "Cannot show authorization notification: POST_NOTIFICATIONS permission not granted")
+        }
+
+        return putResponse.uuid
     }
 
     override fun onStartCommand(
         intent: Intent?,
         flags: Int,
         startId: Int,
     ): Int {
-        startForeground()
-        return START_REDELIVER_INTENT
+        Log.d(logTag, "onStartCommand")
+
+        when (intent?.action) {
```

Comment:
The buttons on a notification will create an `Intent` for this package and class. Now we have 3 intents to handle: "approve", "deny" and "start service", which we distinguish by `intent.action` string.